text/0031-corejam.md

```diff
@@ -0,0 +1,644 @@
+
+Work Items are identified by their Blake2-256 hash, known at the *Item Hash* (`ItemHash`). We provide both the Authorization of the Package and the constituent Work Item Hashes and their Results in order to allow the `refine` logic to take appropriate action in the case that an invalid Work Item was submitted (i.e. one which caused its Refine operation to panic or time-out).
+
+_(Note for later: We may wish to provide a more light-client friendly Work Item identifier than a simple hash; perhaps a Merkle root of equal-size segments.)_
+
+There is an amount of weight which it is allowed to use before being forcibly terminated and any non-committed state changes lost. The lowest amount of weight provided to `accumulate` is defined as the number of `WorkResult` values passed in `results` to `accumulate` multiplied by the `accumulate` field of the Work Class's weight requirements.
+
+However, the actual amount of weight may be substantially more. Each Work Package is allotted a specific amount of weight for all on-chain activity (`weight_per_package` above) and has a weight liability defined by the weight requirements of all Work Items it contains (`total_weight_requirement` above). Any weight remaining after the liability (i.e. `weight_per_package - total_weight_requirement`) may be apportioned to the Work Classes of Items within the Report on a pro-rata basis according to the amount of weight they utilized during `refine`. Any weight unutilized by Classes within one Package may be carried over to the next Package and utilized there.
+
+```rust
+fn get_work_storage(key: &[u8]) -> Option<Vec<u8>>;
+fn get_work_storage_len(key: &[u8]) -> Option<u32>;
+fn checkpoint() -> Weight;
+fn weight_remaining() -> Weight;
+fn set_work_storage(key: &[u8], value: &[u8]) -> Result<(), ()>;
+fn remove_work_storage(key: &[u8]);
+fn ump_enqueue(message: &[u8]) -> Result<(), ()>;
+```
+
+Read-access to the entire Relay-chain state is allowed. No direct write access may be provided since `accumulate` is untrusted code. `set_storage` may fail if an insufficient deposit is held under the Work Class's account.
+
+UMP messages for the Relay-chain to interpret may be enqueued through `ump_enqueue` function. For this to succeed, the Relay-chain must have pre-authorized the use of UMP for this endpoint.
+
+Full access to a child trie specific to the Work Class is provided through the `work_storage` host functions. Since `accumulate` is permissionless and untrusted code, we must ensure that its child trie does not grow to degrade the Relay-chain's overall performance or place untenable requirements on the storage of full-nodes. To this goal, we require an account sovereign to the Work Class to be holding an amount of funds proportional to the overall storage footprint of its Child Trie. `set_work_storage` may return an error should the balance requirement not be met.
```

Comment:
Given that the work class requires some holding of funds to act as an reserve of data, we should support having custom logic to set this up. Like some parachain A may wants to store more data than another parachain B, then parachain B should not be able to use the funds parachain A provided to do this. Bringing me to the conclusion that the `WorkClass` may requires some balance handling logic.

Reply:
Work Class will need to do what makes sense for it, but it should do that internally.

Reply:
Yes, but then we should have some way to call functions in the Work Class from an account?

Reply:
Yes - the permissionless interaction with CoreJam in general needs some detail. I would expect that adding Work Classes be something which is done with a host function available to `accumulate`. Topping up the Work Class's balance would be a similar host function. Tokens would need to be passed between Work Classes for this reason so in lieu of any other solution, we would presumably need a host function for doing this (and that host function could provide a short memo with the transfer to mark the intention).

Reply:
This has now been added as the `transfer` host function